src/dlls/mscoree/unixinterface.cpp

```diff
@@ -146,6 +138,61 @@ int coreclr_initialize(
     ConstWStringArrayHolder propertyValuesW;
     propertyValuesW.Set(StringArrayToUnicode(propertyCount, propertyValues), propertyCount);
 
+    // Process properties that specify CoreCLR startup flags and remove them from the property list
+    STARTUP_FLAGS startupFlags =
+        static_cast<STARTUP_FLAGS>(
+            STARTUP_FLAGS::STARTUP_LOADER_OPTIMIZATION_SINGLE_DOMAIN |
+            STARTUP_FLAGS::STARTUP_SINGLE_APPDOMAIN);
+    int newPropertyIndex = 0;
+    for (int oldPropertyIndex = 0; oldPropertyIndex < propertyCount; ++oldPropertyIndex)
```

Comment:
The code leaks strings of property values / keys that it removes and double deletes property values / keys that it moves down. The StringArrayToUnicode allocates memory for each key / value during the conversion.
I would suggest replacing the StringArrayToUnicode by a function that extracts the new flag properties from the original char* arrays of keys / values and creates the WCHAR* values / keys only for the properties. To simplify things, you can still allocate the WCHAR* arrays with the size equal to the source char* ones and leave number of entries corresponding to the extracted flags set to NULL at the end of the array. The array holder should be ok with it.